search/query.go

```diff
@@ -44,60 +47,150 @@ type PostSearchResult struct {
+	if err := checkParams(offset, size); err != nil {
+		return nil, err
+	}
+	queryStr, filters := ParseQuery(ctx, dir, q)
+	basic := map[string]interface{}{
+		"simple_query_string": map[string]interface{}{
+			"query":            queryStr,
+			"fields":           []string{"everything"},
+			"flags":            "AND|NOT|OR|PHRASE|PRECEDENCE|WHITESPACE",
+			"default_operator": "and",
+			"lenient":          true,
+			"analyze_wildcard": false,
+		},
+	}
+	query := map[string]interface{}{
 		"query": map[string]interface{}{
-			"match": map[string]interface{}{
-				"text": map[string]any{
-					"query":    q,
-					"operator": "and",
+			"bool": map[string]interface{}{
+				"must": basic,
+				"should": []interface{}{
+					map[string]interface{}{"term": map[string]interface{}{"has_avatar": true}},
+					map[string]interface{}{"term": map[string]interface{}{"has_banner": true}},
 				},
+				"filter": filters,
+				"boost":  1.0,
 			},
 		},
 		"size": size,
 		"from": offset,
 	}
 
-	return doSearch(ctx, escli, "posts", query)
+	return doSearch(ctx, escli, index, query)
 }
 
-func doSearchProfiles(ctx context.Context, escli *es.Client, q string) (*EsSearchResponse, error) {
+func DoSearchProfilesTypeahead(ctx context.Context, escli *es.Client, index, q string, size int) (*EsSearchResponse, error) {
+	if err := checkParams(0, size); err != nil {
+		return nil, err
+	}
 	query := map[string]interface{}{
 		"query": map[string]interface{}{
 			"multi_match": map[string]interface{}{
-				"query":    q,
-				"fields":   []string{"description", "displayName", "handle"},
-				"operator": "or",
+				"query": q,
+				"type":  "bool_prefix",
+				"fields": []string{
+					"typeahead",
+					"typeahead._2gram",
+					"typeahead._3gram",
+				},
 			},
 		},
+		"size": size,
 	}
 
-	return doSearch(ctx, escli, "profiles", query)
+	return doSearch(ctx, escli, index, query)
+}
+
+// helper to do a full-featured Lucene query parser (query_string) search, with all possible facets. Not safe to expose publicly.
+func DoSearchGeneric(ctx context.Context, escli *es.Client, index, q string) (*EsSearchResponse, error) {
+	query := map[string]interface{}{
+		"query": map[string]interface{}{
+			"query_string": map[string]interface{}{
+				"query":                  q,
+				"default_operator":       "and",
+				"analyze_wildcard":       true,
+				"allow_leading_wildcard": false,
+				"lenient":                true,
+				"default_field":          "everything",
+			},
+		},
+	}
+
+	return doSearch(ctx, escli, index, query)
 }
 
 func doSearch(ctx context.Context, escli *es.Client, index string, query interface{}) (*EsSearchResponse, error) {
 	var buf bytes.Buffer
 	if err := json.NewEncoder(&buf).Encode(query); err != nil {
-		log.Fatalf("Error encoding query: %s", err)
+		return nil, fmt.Errorf("error encoding query: %s", err)
+	}
+
+	// re-marshal query as string for logging
+	if true {
```

Comment:
Why is this in an `if true`?

Reply:
I can't remember why now... this code block is weird/annoying, will clean it up